File: services/payment-service/infrastructure/correlation_id.py
"""[Issue #8] Correlation ID 검증 헬퍼.

Order Service 가 RabbitMQ 메시지 본문에 넣은 ``correlationId`` 를 그대로 콜백 HTTP
헤더(``X-Correlation-ID``)에 재사용하면, 외부 클라이언트가 주입한 부정 값
(과대길이/제어문자/비ASCII)이 콜백 요청을 깨뜨리고 메시지가 DLQ 로 반복 이동할
수 있다. Boundary 에서 한 번 검증·정규화한다.

규칙:
    - 허용 charset: ``[A-Za-z0-9_-]``
    - 길이: 1 ~ 64 (UUID v4 36자 + 여유)
    - 형식 불일치 시 서버에서 UUID v4 신규 발급
      (메시지 처리 자체는 계속 진행 — 추적 가능성 유지)
"""

from __future__ import annotations

import logging
import re
import uuid

CORRELATION_ID_PATTERN = re.compile(r"^[A-Za-z0-9_-]{1,64}$")

logger = logging.getLogger(__name__)


def normalize_correlation_id(value: object) -> str:
    """검증된 Correlation ID 를 반환한다.

    부정 입력(형식 불일치 문자열, 비문자열)이 들어오면 WARN 로그를 남겨
    공격 시도/포맷 회귀를 운영에서 관측할 수 있게 한다.
    원본 값은 로그에 기록하지 않고(로그 인젝션 방지) 길이/타입만 노출한다.
    누락(``None``/빈 문자열)은 정상 흐름이라 로그를 남기지 않는다.

    Args:
        value: 외부 입력(문자열일 수도, 다른 타입일 수도 있음).

    Returns:
        검증을 통과한 원본 문자열, 또는 새로 발급된 UUID v4 문자열.
    """
    if isinstance(value, str) and CORRELATION_ID_PATTERN.match(value):
        return value

    replacement = str(uuid.uuid4())
    if value not in (None, ""):
        if isinstance(value, str):
            descriptor = f"len={len(value)}"
        else:
            descriptor = f"type={type(value).__name__}"
        logger.warning(
            "correlation_id 부정 입력 치환 — %s → %s",
            descriptor,
            replacement,
        )
    return replacement

File: services/order-service/src/utils/correlationId.js
/**
 * [Issue #8] Correlation ID 검증 헬퍼
 *
 * 외부 클라이언트가 보낸 X-Correlation-ID 값을 그대로 다운스트림(RabbitMQ 메시지,
 * 로그, 콜백 헤더)에 전파하면 악의적 값(과대길이/제어문자/비ASCII)이 들어와
 * Payment Service 의 콜백 헤더 파싱이나 큐 메시지 처리를 망가뜨릴 수 있다
 * (DLQ 오염 가능). Boundary 에서 한 번 검증·정규화한 뒤 사용한다.
 *
 * 규칙:
 *   - 허용 charset: [A-Za-z0-9_-] (URL-safe 문자만)
 *   - 길이: 1 ~ 64 (UUID v4 = 36자 + 여유)
 *   - 그 외(비문자열/형식 불일치)면 서버에서 UUID 를 새로 발급한다.
 *     요청을 거부하지 않는 이유: 추적 가능성을 항상 유지하기 위함.
 */
const { randomUUID } = require('crypto');

const CORRELATION_ID_PATTERN = /^[A-Za-z0-9_-]{1,64}$/;

/**
 * 검증/정규화된 Correlation ID 를 반환한다.
 *
 * 부정 입력(형식 불일치 문자열, 비문자열)이 들어오면 WARN 로그를 남겨
 * 공격 시도/포맷 회귀를 운영에서 관측할 수 있게 한다.
 * 원본 값은 로그에 찍지 않는다 — 로그 인젝션 방지를 위해 길이/타입만 노출.
 * 누락(undefined/null/빈 문자열)은 정상 흐름이라 로그 생략.
 *
 * @param {unknown} value 외부에서 전달된 원시 값 (헤더, 메시지 본문 등)
 * @returns {string}
 */
function normalizeCorrelationId(value) {
    if (typeof value === 'string' && CORRELATION_ID_PATTERN.test(value)) {
        return value;
    }
    const replacement = randomUUID();
    if (value !== undefined && value !== null && value !== '') {
        const descriptor = typeof value === 'string'
            ? `len=${value.length}`
            : `type=${typeof value}`;
        console.warn(
            `[correlation-id] 부정 입력 치환 — ${descriptor} → ${replacement}`
        );
    }
    return replacement;
}

module.exports = { normalizeCorrelationId, CORRELATION_ID_PATTERN };
